src/Microsoft.AspNet.Authentication.Cookies/Events/CookieAuthenticationEvents.cs

```diff
@@ -71,24 +94,73 @@ public class CookieAuthenticationEvents : ICookieAuthenticationEvents
         /// Implements the interface method by invoking the related delegate method.
         /// </summary>
         /// <param name="context">Contains information about the event</param>
-        public virtual Task RedirectToLogout(CookieRedirectContext context) => OnRedirect(context);
+        public virtual Task RedirectToLogout(CookieRedirectContext context)
+        {
+            if (IsAjaxRequest(context.Request))
+            {
+                context.Response.Headers["Location"] = context.RedirectUri;
+                context.Response.StatusCode = 200;
+                return Task.FromResult(0);
+            }
+            else
+            {
+                return OnRedirect(context);
+            }
+        }
 
         /// <summary>
         /// Implements the interface method by invoking the related delegate method.
         /// </summary>
         /// <param name="context">Contains information about the event</param>
-        public virtual Task RedirectToLogin(CookieRedirectContext context) => OnRedirect(context);
+        public virtual Task RedirectToLogin(CookieRedirectContext context)
+        {
+            if (IsAjaxRequest(context.Request))
```

Comment:
Really? Then it was a terrible idea, nobody uses these directly. Maybe I meant that the same func could be assigned to each of them.